### PR TITLE
fix: seated event custom questions

### LINKED ISSUE #12188: [CAL-2906] Event with seating: The custom question answers of the first user to book are unintentionally exposed to subsequent attendees when rescheduling or canceling.
### Issue Summary

When you host an event that provides seats and includes custom booking questions, the answers of the first user to book are inadvertently revealed to subsequent attendees of the event when they attempt to reschedule or cancel.

### Steps to Reproduce

1. Create an event that offers seats, uncheck both checkboxes 'Share attendee information between guests' and 'Show the number of available seats.'
2. Add custom questions like phone number, zip code, and so on.
3. Register for this event.
4. Register again for the same event on the same day and time, but provide different answers and use a different email.
5. Open the second registrant's email inbox for the confirmation email and click on the button to cancel or reschedule the meeting.
6. Notice that the first person to register for this event appears in the field instead of the user's own questions.

### Actual Results

* The answers of the first user to book are inadvertently revealed to subsequent attendees of the event when they attempt to reschedule or cancel.

### Expected Results

* To display the answers of the correct attendee instead of the first attendee, or have them appear blank instead.

### Technical details

* Production env.

### Evidence

image

Screenshot 2023-11-01 at 17 02 33

Screenshot 2023-11-01 at 17 07 53

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2906](https://linear.app/calcom/issue/CAL-2906/event-with-seating-the-custom-question-answers-of-the-first-user-to)</sub>

### PR DESCRIPTION
**Pull request by MehulZR**
## What does this PR do?

Fixes #12188 (issue)

https://github.com/calcom/cal.com/assets/86384652/7fd10db2-d0c8-4071-a2f0-bbf533c2678a

https://github.com/calcom/cal.com/assets/86384652/292ec872-af1e-4281-bb85-ec7d6357ee64

Also previously ff `share attendee info` was false and a non-host user visits the booking page without `seatReferenceId`. There was a leak of info for the first user.

Before:
Screenshot from 2023-12-13 20-28-17
After:
Screenshot from 2023-12-13 20-19-58

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

**Review by PeerRich — DISMISSED** [review]
we should not be showing "Nameless" anywhere in the UI.

i know we do it somewhere when there is no username but ideally we get rid of it

**Comment by MehulZR** [comment]
Explaination

**Comment by MehulZR** [comment]
Explaination

E2E 3  are having issue, because the seats are being created without the responses basic values
E2E 7,9,10 are passing on my side, dont know why they are failing in the PR ( probably the nextjs server crashing? recently has been happening with me and contributors on discord too. Seems to be a heap memory issue. Trying to look into it. )

**Comment by CarinaWolli** [comment]
> E2E 3 are having issue, because the seats are being created without the responses basic values

If that's intended, can you fix the tests so they pass? 

> E2E 7,9,10 are passing on my side, dont know why they are failing in the PR ( probably the nextjs server crashing? recently has been happening with me and contributors on discord too. Seems to be a heap memory issue. Trying to look into it. )

Could have been flaky tests. We are not running e2e tests on every PR anymore, but please make sure that `booking-seats.e2e.ts` passes locally

**Comment by MehulZR** [comment]
> If that's intended, can you fix the tests so they pass?
>
> Could have been flaky tests. We are not running e2e tests on every PR anymore, but please make sure that `booking-seats.e2e.ts` passes locally

I have updated the tests & E2E 3, 7, 9, 10 are working correctly now. 
Was not able to make the Stripe E2E tests run, So can we run the E2E test on this PR?
It will be great if we can test that before merging this.

File: apps/web/pages/booking/[uid].tsx [hunk 1]
```diff
@@ -109,35 +120,30 @@ export default function Success(props: SuccessProps) {
...
-  const cancellationReason = props.bookingInfo.cancellationReason || props.bookingInfo.rejectionReason;
+  const status = bookingInfo?.status;
+  const reschedule = bookingInfo.status === BookingStatus.ACCEPTED;
+  const cancellationReason = bookingInfo.cancellationReason || bookingInfo.rejectionReason;
 
-  const attendeeName =
-    typeof props?.bookingInfo?.attendees?.[0]?.name === "string"
-      ? props?.bookingInfo?.attendees?.[0]?.name
-      : "Nameless";
+  const attendeeName = parseName(bookingInfo.responses.name);
```
Comment (joeauyeung) [inline apps/web/pages/booking/[uid].tsx:144]: I love setting the scheduler's name dynamically on the booking page. Although not every event type will have that field in the booking title. Could we move this into the `getEventName` and only parse for the name when `{SCHEDULER}` is a part of the booking title?

**Review by joeauyeung — DISMISSED** [review]
Great work so far @MehulZR! 

I left one piece of feedback below. Like Carina mentioned just ensure that tests are passing locally as well.

File: apps/web/pages/booking/[uid].tsx (hunk 1 above)
Comment (MehulZR) [reply apps/web/pages/booking/[uid].tsx:144]: Hey @joeauyeung, I am afraid I don't understand you completely.

Comment (joeauyeung) [reply apps/web/pages/booking/[uid].tsx]: When setting up event types, you can have the name of the event be the `attendeeName`. I think we should only parse for the attendee name if the `{SCHEDULER}` place holder is a part of the event name.
CleanShot 2024-01-12 at 13 27 28

In terms of the Stripe tests what error are you getting exactly? One thing that's not a part of that doc is locally you need to send webhook events to your local machine. You can use the Stripe CLI and use the `stripe listen --forward to localhost:3000/api/integrations/stripepayment/webhook`

Comment (MehulZR) [reply apps/web/pages/booking/[uid].tsx]: Thanks for the elaboration, I will make the necessary changes tomorrow. 

Last time I tried forwarding to `localhost:3000/api/integrations/stripepayment/webhook` but was getting an error `Not able to post` from the stripe cli. I was using the stripe cli from docker.

Will try to run it again by native package tomorrow.

Comment (MehulZR) [reply apps/web/pages/booking/[uid].tsx]: I was able to get past the `Not able to post` error by using the native stripe cli. However the test aren't passing even on `Main` branch for me. The steps & assertions in the tests seems to work as far as I understand. But end up getting an Aggregate Error. 

image

Tried digging in it, but the error is bit ambiguous. On the development server output, All I get is an 
`Webhook Error: Unhandled Stripe Webhook event type { stripe event type such as account updated/ authorized }`. Which comes from [here](https://github.com/calcom/cal.com/blob/e5eb121ab75fdf0d5b5ffa1b4fef61107329e4cf/packages/features/ee/payments/api/webhook.ts#L156C1-L159C10) and is expected I believe.

Aside from all this, I see nothing suspicious that I can look into.

**Comment by MehulZR** [comment]
While trying to fix types i accidentally broke the functionality. Reverting it back for the time being.

**Comment by MehulZR** [comment]
> While trying to fix types i accidentally broke the functionality. Reverting it back for the time being.

resolved

**Comment by keithwillcode** [comment]
@MehulZR looks like there are conflicts

**Comment by MehulZR** [comment]
> @MehulZR looks like there are conflicts

Thanks for the heads-up, I am working on an another PR, will work on this one soon.

File: packages/features/bookings/lib/get-booking.ts [hunk 2]
```diff
@@ -185,10 +191,14 @@ export const getBookingForReschedule = async (uid: string, userId?: number) => {
...
   if (!theBooking && !rescheduleUid) return null;
 
-  const booking = await getBooking(prisma, rescheduleUid || uid);
+  const booking = await getBooking(prisma, rescheduleUid || uid, bookingSeatReferenceUid ? true : false);
 
   if (!booking) return null;
 
+  if (bookingSeatReferenceUid) {
+    booking["description"] = bookingSeatData?.description ?? null;
+    booking["responses"] = bookingResponsesDbSchema.parse(bookingSeatData?.responses);
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/get-booking.ts:200]:
```suggestion
    booking["responses"] = bookingResponsesDbSchema.parse(bookingSeatData?.responses ?? {});
```
same here

**Review by CarinaWolli — CHANGES_REQUESTED** [review]

**Review by CarinaWolli — DISMISSED** [review]
Tested it and everything looks good now 👍 Great work @MehulZR 👏🏻

**Comment by keithwillcode** [comment]
@MehulZR converting to draft while merge conflicts are fixed. Feel free to mark as ready review again afterwards. 

/bonus 20

**Review by CarinaWolli — APPROVED** [review]

**Pull request merged**